File: artifacts/api-server/src/routes/predictions.ts
import { Router, type IRouter } from "express";
import { eq, and } from "drizzle-orm";
import { db, gradesTable, assignmentsTable, enrollmentsTable, studentsTable, coursesTable } from "@workspace/db";
import { PredictGradeBody, PredictAtRiskParams } from "@workspace/api-zod";
import { scoreToLetter, toPercent, riskLevel } from "../lib/gradeUtils";

const router: IRouter = Router();

/**
 * Simple linear-weighted grade prediction model.
 * Weights: midterm (35%), assignment avg (45%), completion rate (20%)
 */
function predictFinalScore(
  midtermPct: number | null,
  assignmentAvgPct: number | null,
  completionRate: number, // 0-1
): { score: number; confidence: number } {
  const components: { value: number; weight: number }[] = [];

  if (midtermPct !== null) components.push({ value: midtermPct, weight: 0.35 });
  if (assignmentAvgPct !== null) components.push({ value: assignmentAvgPct, weight: 0.45 });
  components.push({ value: completionRate * 100, weight: 0.20 });

  const totalWeight = components.reduce((s, c) => s + c.weight, 0);
  const score = components.reduce((s, c) => s + c.value * (c.weight / totalWeight), 0);
  // Confidence based on how much data we have
  const confidence = Math.min(100, Math.round(totalWeight * 90 + completionRate * 10));

  return { score: Math.max(0, Math.min(100, score)), confidence };
}

// POST /predictions/grade
router.post("/predictions/grade", async (req, res): Promise<void> => {
  const parsed = PredictGradeBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const d = parsed.data as {
    student_id: number;
    course_id: number;
    midterm_score?: number | null;
    assignment_completion_rate?: number | null;
  };

  // Fetch actual grades for this student/course
  const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, d.course_id));

  let midtermPct: number | null = d.midterm_score ?? null;
  let assignmentTotal = 0;
  let assignmentCount = 0;
  let submittedCount = 0;

  for (const asgn of assignments) {
    const grade = await db.select().from(gradesTable).where(and(eq(gradesTable.studentId, d.student_id), eq(gradesTable.assignmentId, asgn.id))).limit(1);
    if (grade[0]) {
      const pct = toPercent(Number(grade[0].score), Number(asgn.maxScore));
      if (asgn.type === "midterm" && midtermPct === null) midtermPct = pct;
      else {
        assignmentTotal += pct;
        assignmentCount++;
      }
      submittedCount++;
    }
  }

  const completionRate = d.assignment_completion_rate ?? (assignments.length > 0 ? submittedCount / assignments.length : 0);
  const assignmentAvg = assignmentCount > 0 ? assignmentTotal / assignmentCount : null;

  const { score, confidence } = predictFinalScore(midtermPct, assignmentAvg, completionRate);
  const letter = scoreToLetter(score);

  const factors = [
    { factor: "Midterm Performance", weight: 0.35, value: midtermPct ?? 0 },
    { factor: "Assignment Average", weight: 0.45, value: assignmentAvg ?? 0 },
    { factor: "Completion Rate", weight: 0.20, value: completionRate * 100 },
  ];

  res.json({
    student_id: d.student_id,
    course_id: d.course_id,
    predicted_score: Math.round(score * 10) / 10,
    predicted_letter: letter,
    confidence,
    risk_level: riskLevel(score),
    factors,
  });
});

// GET /predictions/at-risk/:courseId
router.get("/predictions/at-risk/:courseId", async (req, res): Promise<void> => {
  const params = PredictAtRiskParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const courseId = params.data.courseId;
  const enrollments = await db.select().from(enrollmentsTable).where(eq(enrollmentsTable.courseId, courseId));
  const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, courseId));

  const result = [];

  for (const enr of enrollments) {
    const student = await db.select().from(studentsTable).where(eq(studentsTable.id, enr.studentId)).limit(1);
    if (!student[0]) continue;

    let midtermPct: number | null = null;
    let assignmentTotal = 0;
    let assignmentCount = 0;
    let submittedCount = 0;
    let currentWeightedScore = 0;
    let currentTotalWeight = 0;

    for (const asgn of assignments) {
      const grade = await db.select().from(gradesTable).where(and(eq(gradesTable.studentId, enr.studentId), eq(gradesTable.assignmentId, asgn.id))).limit(1);
      if (grade[0]) {
        const pct = toPercent(Number(grade[0].score), Number(asgn.maxScore));
        if (asgn.type === "midterm" && midtermPct === null) midtermPct = pct;
        else { assignmentTotal += pct; assignmentCount++; }
        currentWeightedScore += pct * Number(asgn.weight);
        currentTotalWeight += Number(asgn.weight);
        submittedCount++;
      }
    }

    const completionRate = assignments.length > 0 ? submittedCount / assignments.length : 0;
    const assignmentAvg = assignmentCount > 0 ? assignmentTotal / assignmentCount : null;
    const { score, confidence } = predictFinalScore(midtermPct, assignmentAvg, completionRate);

    const currentScore = currentTotalWeight > 0 ? Math.round((currentWeightedScore / currentTotalWeight) * 10) / 10 : null;

    result.push({
      student_id: student[0].id,
      student_name: student[0].name,
      predicted_score: Math.round(score * 10) / 10,
      predicted_letter: scoreToLetter(score),
      risk_level: riskLevel(score),
      confidence,
      current_score: currentScore,
    });
  }

  // Sort by predicted score ascending (most at-risk first)
  result.sort((a, b) => a.predicted_score - b.predicted_score);

  res.json(result);
});

export default router;

File: artifacts/api-server/src/routes/predictions.py
"""Prediction routes — grade prediction + at-risk by course."""
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.db.session import get_db
from src.db.models import AssignmentRow, GradeRow, EnrollmentRow, StudentRow
from src.domain.grade_utils import to_percent, score_to_letter, risk_level

router = APIRouter()


def predict_final_score(midterm_pct, assignment_avg_pct, completion_rate: float):
    """Linear-weighted grade prediction: midterm 35%, assignments 45%, completion 20%."""
    components = []
    if midterm_pct is not None:
        components.append({"value": midterm_pct, "weight": 0.35})
    if assignment_avg_pct is not None:
        components.append({"value": assignment_avg_pct, "weight": 0.45})
    components.append({"value": completion_rate * 100, "weight": 0.20})
    total_weight = sum(c["weight"] for c in components)
    score = sum(c["value"] * (c["weight"] / total_weight) for c in components)
    confidence = min(100, round(total_weight * 90 + completion_rate * 10))
    return {"score": max(0.0, min(100.0, score)), "confidence": confidence}


class PredictGradeBody(BaseModel):
    student_id: int
    course_id: int
    midterm_score: Optional[float] = None
    assignment_completion_rate: Optional[float] = None


@router.post("/predictions/grade")
def predict_grade(body: PredictGradeBody, db: Session = Depends(get_db)):
    assignments = db.query(AssignmentRow).filter(AssignmentRow.course_id == body.course_id).all()
    midterm_pct = body.midterm_score
    assignment_total = 0.0
    assignment_count = 0
    submitted_count = 0
    for asgn in assignments:
        grade = db.query(GradeRow).filter(
            GradeRow.student_id == body.student_id,
            GradeRow.assignment_id == asgn.id,
        ).first()
        if grade:
            pct = to_percent(float(grade.score), float(asgn.max_score))
            if asgn.type == "midterm" and midterm_pct is None:
                midterm_pct = pct
            else:
                assignment_total += pct
                assignment_count += 1
            submitted_count += 1
    completion_rate = (
        body.assignment_completion_rate
        if body.assignment_completion_rate is not None
        else (submitted_count / len(assignments) if assignments else 0)
    )
    assignment_avg = assignment_total / assignment_count if assignment_count > 0 else None
    pred = predict_final_score(midterm_pct, assignment_avg, completion_rate)
    score = pred["score"]
    letter = score_to_letter(score)
    factors = [
        {"factor": "Midterm Performance", "weight": 0.35, "value": midterm_pct or 0},
        {"factor": "Assignment Average", "weight": 0.45, "value": assignment_avg or 0},
        {"factor": "Completion Rate", "weight": 0.20, "value": completion_rate * 100},
    ]
    return {
        "student_id": body.student_id,
        "course_id": body.course_id,
        "predicted_score": round(score * 10) / 10,
        "predicted_letter": letter,
        "confidence": pred["confidence"],
        "risk_level": risk_level(score),
        "factors": factors,
    }


@router.get("/predictions/at-risk/{course_id}")
def at_risk_by_course(course_id: int, db: Session = Depends(get_db)):
    enrollments = db.query(EnrollmentRow).filter(EnrollmentRow.course_id == course_id).all()
    assignments = db.query(AssignmentRow).filter(AssignmentRow.course_id == course_id).all()
    result = []
    for enr in enrollments:
        student = db.query(StudentRow).filter(StudentRow.id == enr.student_id).first()
        if not student:
            continue
        midterm_pct = None
        assignment_total = 0.0
        assignment_count = 0
        submitted_count = 0
        current_weighted = 0.0
        current_total_weight = 0.0
        for asgn in assignments:
            grade = db.query(GradeRow).filter(
                GradeRow.student_id == enr.student_id,
                GradeRow.assignment_id == asgn.id,
            ).first()
            if grade:
                pct = to_percent(float(grade.score), float(asgn.max_score))
                if asgn.type == "midterm" and midterm_pct is None:
                    midterm_pct = pct
                else:
                    assignment_total += pct
                    assignment_count += 1
                current_weighted += pct * float(asgn.weight)
                current_total_weight += float(asgn.weight)
                submitted_count += 1
        completion_rate = submitted_count / len(assignments) if assignments else 0
        assignment_avg = assignment_total / assignment_count if assignment_count > 0 else None
        pred = predict_final_score(midterm_pct, assignment_avg, completion_rate)
        current_score = round(current_weighted / current_total_weight * 10) / 10 if current_total_weight > 0 else None
        result.append({
            "student_id": student.id,
            "student_name": student.name,
            "predicted_score": round(pred["score"] * 10) / 10,
            "predicted_letter": score_to_letter(pred["score"]),
            "risk_level": risk_level(pred["score"]),
            "confidence": pred["confidence"],
            "current_score": current_score,
        })
    result.sort(key=lambda x: x["predicted_score"])
    return result
